### ISSUE TITLE
Warn/error on "over ... with this"

### ISSUE THREAD

**Issue by philrz** [brimdata/super#5215]
## tl;dr

A community user recently found the Zed language allowed them to write a lateral subquery preceded by `over ... with this` and the resulting overlapping uses of `this` values inside the subquery caused confusion. Since a lateral subquery is often used to work with the sequence of outer values via inner references to `this`, we recognized it would be helpful to surface a warning or error if a user's program invokes this overlapping usage, since it's probably a mistake.

## Details

Repro is with Zed commit 71e35c5.

This issue was surfaced in a [community Slack thread](https://brimdata.slack.com/archives/CTSMAK6G7/p1723411333396999). They first started with the query shown below, which I'm executing now with our own simplified [aws.json](https://github.com/user-attachments/files/16591287/aws.json) test data:

```
{
    "Reservations": [
        {
            "Groups": [],
            "Instances": [
                {
                    "InstanceId": "i-033bd8fe8b30bb00f",
                    "Tags": [
                        {
                            "Key": "Name",
                            "Value": "shasta"
                        }
                    ]
                }
            ]
        },
        {
            "Groups": [],
            "Instances": [
                {
                    "InstanceId": "i-019bcf4e640743aa6",
                    "Tags": [
                        {
                            "Key": "Name",
                            "Value": "phil-win2019"
                        }
                    ]
                }
            ]
        }
    ]
}
```

```
$ zq -version
Version: v1.17.0-20-g71e35c5d

$ zq -z 'over Reservations | over Instances
  | over Tags with obj=this => ( where Key=="Name" | {name: this.Value, id:obj.InstanceId} )' aws.json
{name:"shasta",id:"i-033bd8fe8b30bb00f"}
{name:"phil-win2019",id:"i-019bcf4e640743aa6"}
```

The user then remarked:

> Huh - I noticed with a couple small edits - I could say `with this` (vs. `with obj=this`) in the lateral subquery ... and by removing the `this` from `this.Value` since those collide now ... it still does work:
>
>```
>zq -z '
>  over Reservations | over Instances
>  | over Tags with this => ( where Key=="Name" | {name: Value, id:this.InstanceId} )' aws.json
>{name:"shasta",id:"i-033bd8fe8b30bb00f"}
>{name:"phil-win2019",id:"i-019bcf4e640743aa6"}
>```
>
> But - that feels ... dangerous? Or odd or something? Maybe it's not.

The way he was making an implied field reference to just `Value` was allowing him to still access the "inner" data that would normally be referenced as `this.Value`. He ultimately figured out on his own what was going on.

> I guess it just means that `this` is going to stomp the "implied" `this` inside the subquery. that's at least what it does:
>```
>zq -z '
>  over Reservations | over Instances
>  | over Tags with this => ( where Key=="Name" | {name: this.Value, id:this.InstanceId} )' aws.json
>{name:error("missing"),id:"i-033bd8fe8b30bb00f"}
>{name:error("missing"),id:"i-019bcf4e640743aa6"}
>```

However, since it required a close reading of the docs to clear up the confusion, some messaging from the tooling would have also been helpful. In the user's opinion:

> off the cuff, i’d vote for warning rather than error … it’s not broken per se, just … stompy :hiking_boot: :smile:

**Comment by philrz** [comment]
With the changes from `over` to `unnest` in #6021, `with` and the concept of variables are gone, so it's no longer possible to bump into this issue.

The [draft `unnest` docs](https://superdb.org/book/super-sql/operators/unnest.html) cover some of the new functionality and there's more info to come. But here's an example of how to convert the query above that previously caused trouble to current syntax:

```
$ super -version
Version: e214c5f08

$ super -s -c '
unnest Reservations
| unnest Instances
| unnest {InstanceId,Tag:Tags} into (
    where Tag.Key=="Name"
    | values {name:Tag.Value, id:InstanceId}
)' aws.json

{name:"shasta",id:"i-033bd8fe8b30bb00f"}
{name:"phil-win2019",id:"i-019bcf4e640743aa6"}
```

In terms of the "stomping" that was previously possisble, it's true that one could still use the word `this` in ways that could be confusing, e.g., compare these two scaled back variations on this query:

```
$ super -s -c '
unnest Reservations
| unnest Instances
| unnest {InstanceId,Tag:Tags} into (
  values this
)' aws.json

{InstanceId:"i-033bd8fe8b30bb00f",Tag:{Key:"Name",Value:"shasta"}}
{InstanceId:"i-019bcf4e640743aa6",Tag:{Key:"Name",Value:"phil-win2019"}}

$ super -s -c '
unnest Reservations
| unnest Instances
| unnest {this:InstanceId,Tag:Tags} into (
  values this
)' aws.json

{this:"i-033bd8fe8b30bb00f",Tag:{Key:"Name",Value:"shasta"}}
{this:"i-019bcf4e640743aa6",Tag:{Key:"Name",Value:"phil-win2019"}}
```

But all we're really seeing here is just ordinary "working with records" stuff. The tools make no attempt to prevent use of `this` as a record key, e.g., compare:

```
$ echo '{foo: "bar"}' | super -c 'values this' -
{foo:"bar"}

$ echo '{foo: "bar"}' | super -c 'values this.foo' -
"bar"

$ echo '{foo: "bar"}' | super -c 'values foo' -
"bar"

$ echo '{this: "bar"}' | super -c 'values this' -
{this:"bar"}

$ echo '{this: "bar"}' | super -c 'values this.this' -
"bar"
```

i.e., if you're in the unfortunate spot of having `this` as a record key, because of ambiguity you can't use the shortcut of getting its value by referencing _just_ `this` like we did with _just_ `foo`, but one can always be explicit with `this.this` just like we were with `this.foo`.

@chrismo: I know this was originally one of your issues and that you've been one of the few users of `over` and lateral subqueries in your real world use cases, so feel free to reach out if you need any assistance getting the hang of the new stuff, since it's all pretty fresh. If you're not in a hurry the docs will be improving very soon, so you can wait for that and try self-service as well. Up to you!

Closing this issue since it's no longer an exposure.
